cpp/src/arrow/compute/kernels/vector_nested.cc

```diff
@@ -76,6 +76,24 @@ struct ListParentIndicesArray {
 
   Status Visit(const LargeListType& type) { return VisitList(type); }
 
+  Status Visit(const FixedSizeListType& type) {
+    const int32_t slot_length = type.list_size();
+    const int64_t values_length = slot_length * (input->length - input->GetNullCount());
+    ARROW_ASSIGN_OR_RAISE(auto indices, ctx->Allocate(values_length * sizeof(int32_t)));
+    int32_t* out_indices = reinterpret_cast<int32_t*>(indices->mutable_data());
+    const uint8_t* bitmap = input->GetValues<uint8_t>(0, 0);
+    for (int32_t i = 0; i < input->length; i++) {
+      if (!bitmap || BitUtil::GetBit(bitmap, input->offset + i)) {
+        std::fill(out_indices, out_indices + slot_length,
```

Comment:
Move the `if (!bitmap)` check outside of the loop.

Reply:
Also, you can probably track consecutive valid bits, and perform a `std::fill` when an invalid bit is encountered.

Reply:
You can't batch copies because different slots have different values.

Reply:
Also I would expect the branch predictor to make the difference moot here anyways; if we move `!bitmap` outside the loop, we would have to add a separate loop to fill the output array.

Reply:
Also, if you wanted to do that, it would be much better to call `VisitSetBitRuns`...